### ISSUE TITLE
 IOU – Not here page when create new request after canceling payment for the first one

### ISSUE THREAD

**Issue by lanitochka17** [Expensify/App#49801]
If you haven’t already, check out our [contributing guidelines](https://github.com/Expensify/ReactNativeChat/blob/main/contributingGuides/CONTRIBUTING.md) for onboarding and email contributors@expensify.com to request to join our Slack channel!
___

**Version Number:** 9.0.40-0
**Reproducible in staging?:** Y
**Reproducible in production?:** Y
**If this was caught during regression testing, add the test name, ID and link from TestRail:** N/A
**Email or phone of affected tester (no customers):** gocemate+a2282@gmail.com
**Issue reported by:** Applause - Internal Team

## Action Performed:
1. Go to workspace chat> + > Submit expense> Complete the flow
2. Pay the request
3. Go to transaction thread> Click on header> Cancel payment
4. Go to main page> + > Submit expense> Complete the flow
5. Navigate to newly created request

## Expected Result:
Transaction thread should open

## Actual Result:
Not here page appears

## Workaround:
Unknown

## Platforms:

Which of our officially supported platforms is this issue occurring on?
- [ ] Android: Native
- [ ] Android: mWeb Chrome
- [ ] iOS: Native
- [ ] iOS: mWeb Safari
- [ ] MacOS: Chrome / Safari
- [ ] MacOS: Desktop

## Screenshots/Videos

Add any screenshot/video evidence

</details> 

https://github.com/user-attachments/assets/79bdf5a8-0aca-4006-b29c-98923dce8ecf

[View all open jobs on GitHub](https://github.com/Expensify/App/issues?q=is%3Aopen+is%3Aissue+label%3A%22Help+Wanted%22)

**Comment by lanitochka17** [comment]
@sonialiap  FYI I haven't added the External label as I wasn't 100% sure about this issue. Please take a look and add the label if you agree it's a bug and can be handled by external contributors

**Comment by nkdengineer** [comment]
## Proposal

### Please re-state the problem that we are trying to solve in this issue.
Not here page appears

### What is the root cause of that problem?
- When we pay money requests, the `iouReportID` of the chat report is updated to `null`. But it's not updated to the expense report again when we cancel the payment. 

- Then when we submitted another request, we created a new optimistic report preview action and a new expense report but the backend moved this to the canceled payment expense report. So not found page appears when we open the expense report from the optimistic report preview action.

### What changes do you think we should make in order to solve the problem?
When we cancel payment here, we should update `iouReportID` in optimistic data here if `chatReport.iouReportID` is `null` or we can check if the current `chatReport.iouReportID` is created before or after the `expenseReport` to decide update `iouReportID` or not.  

```
{
    onyxMethod: Onyx.METHOD.MERGE,
    key: `${ONYXKEYS.COLLECTION.REPORT}${chatReport.reportID}`,
    value: {
        iouReportID: chatReport?.iouReportID ? chatReport?.iouReportID : expenseReport.reportID,
    },
}
```
https://github.com/Expensify/App/blob/70ed492da2743309039e17827d59792e6a148310/src/libs/actions/IOU.ts#L7430

We also need a BE change to update `iouReportID` of the chat report when we cancel the payment because

- `iouReportID` isn't updated after we cancel the payment
- If we submit another expense after canceling payment, BE still moves this to the canceled expense report

### What alternative solutions did you explore? (Optional)
NA

**Comment by rushatgabhane** [comment]
that makes sense to me. 🎀 👀 🎀 

https://github.com/Expensify/App/issues/49801#issuecomment-2378861880

**Comment by rushatgabhane** [comment]
> the iouReportID of the chat report is updated to null. But it's not updated to the expense report again when we cancel the payment.

@nkdengineer how will we know the old `iouReportID` that we need to optimistically reset to after cancellling payment?

**Comment by nkdengineer** [comment]
> @nkdengineer how will we know the old iouReportID that we need to optimistically reset to after cancellling payment?

@rushatgabhane When we create a money request in the `chatReport` (DM or policy expense chat), we get the IOU report via `chatReport.iouReportID`. So we need to update `iouReportID` of the `chatReport`.

https://github.com/Expensify/App/blob/69483ff0ef36e39fbf40c15f2caa3471dccbef17/src/libs/actions/IOU.ts#L2053-L2056

> what will happen when a report has multiple expenses and we cancel only one of them?

We can check if the current `iouReportID` of the `chatReport` is created before the cancelling iou report or `iouReportID` we need to update `iouReportID` to the cancelling iou report otherwise we can keep the current `chatReport.iouReportID`

**Comment by nkdengineer** [comment]
@tgolen Can you review the comment [here](https://github.com/Expensify/App/issues/49801#issuecomment-2384679358)?

**Comment by tgolen** [comment]
Hm, yeah... sorry. I remember seeing that, and I was expecting @rushatgabhane to respond to that before I assigned anything. I've looked at the comment, and the answer seems to make sense to me, but I am not very familiar with the code or the flow. @nkdengineer do you think that's just something we can tackle in the PR? If so, I'll go ahead and assign you.

**Comment by nkdengineer** [comment]
@tgolen Based on this [comment](https://github.com/Expensify/App/issues/49801#issuecomment-2384733514), can you implement the BE change to return the correct `iouReportID` of the chat report after we cancel payment? Thanks.

**Comment by tgolen** [comment]
Yeah, I can do that. However, I'm not entirely sure what needs to be changed. Can you please detail the backend problem a little more with an expected/actual scenario and reproduction steps?

**Comment by nkdengineer** [comment]
@tgolen I just checked again, and after resetting the cache and restarting, the `iouReportID` of the chat report is updated correctly now. We only need to fix the optimistic data case in the front end now. 

OPTIONAL: But I prefer if we can return the `iouReportID` data of chat report when we call `CancelPayment` API.

**Comment by nkdengineer** [comment]
@rushatgabhane The PR is ready.

**Comment by tgolen** [comment]
@nkdengineer bump on my question [here](https://github.com/Expensify/App/issues/49801#issuecomment-2417181770) about the BE fix you're looking for.

**Comment by nkdengineer** [comment]
@tgolen I answered [here](https://github.com/Expensify/App/issues/49801#issuecomment-2418382381).

**Comment by tgolen** [comment]
Oh, hahaha. Sorry I walked right by that somehow.

> I prefer if we can return the iouReportID data of chat report when we call CancelPayment API.

I can help with this if it would make the API more consistent and predictable. How would you want it to be returned? I assume it would just be the `iouReportID` property for the key `report_<expenseReportID>`?

**Comment by nkdengineer** [comment]
> I can help with this if it would make the API more consistent and predictable. How would you want it to be returned? I assume it would just be the iouReportID property for the key report_<expenseReportID>?

@tgolen  It should be `report_<chatReportID>`

**Comment by tgolen** [comment]
OK, got it. I'll create an issue for that and I can let you know when it's deployed if you'd like.

**Comment by nkdengineer** [comment]
Thanks.

**Comment by tgolen** [comment]
I just submitted a PR for it, so it should get merged and deployed early next week. I'll let you know when it does.

**Comment by sonialiap** [comment]
"Include the iouReportID of the chat report in the CancelPayment response" deployed to production 19 hours ago

**Comment by tgolen** [comment]
Yes, the backend things are completed on this. What's the next step for this issue?

**Comment by nkdengineer** [comment]
Waiting update from @rushatgabhane.

**Comment by sonialiap** [comment]
Update - PR in process https://github.com/Expensify/App/pull/50847

**Comment by sonialiap** [comment]
I'm OOO Dec 16-20, adding a leave buddy

next steps:
- @rushatgabhane to complete [checklist](https://github.com/Expensify/App/issues/49801#issuecomment-2537549996)
- BZ to make payment

**Comment by garrettmknight** [comment]
$250 approved for @rushatgabhane

**Comment by sonialiap** [comment]
Payment summary:
@rushatgabhane $250 via NewDot
@nkdengineer $250 paid in upwork ✔️
